charmcraft/utils.py

```diff
@@ -405,3 +409,67 @@ def get_charm_name_from_path(path: pathlib.Path) -> str:
     if metadata_yaml is None or "name" not in metadata_yaml:
         raise InvalidCharmPathError(path)
     return metadata_yaml["name"]
+
+
+def get_pypi_packages(*requirements: Iterable[str]) -> Set[str]:
+    packages = set()
+    for req in requirements:
+        for line in req:
+            line = line.strip()
+            if line.startswith("-"):
+                continue
+            if match := PACKAGE_LINE_REGEX.match(line):
+                packages.add(match.group(0))
+
+    return packages
+
+
+def exclude_packages(requirements: Set[str], *, excluded: Collection[str]) -> Set[str]:
+    exclusions = set()
+    for requirement in requirements:
+        if match := PACKAGE_LINE_REGEX.match(requirement):
+            if match.group(0) in excluded:
+                exclusions.add(requirement)
+
+    return requirements - exclusions
+
+
+def get_pip_command(
```

Comment:
How does this work with hashed requirement files? Or with a mix of hashed and unhashed requirement files?

Reply:
It doesn't work with a mix (current behaviour falls back to the charmcraft 2.3 behaviour), but it does work with hashed files.